flaskdeid/annotation.py

```diff
@@ -105,26 +118,54 @@ def source_origins(self):
 
     @property
     def type(self):
-        if len(set(self.source_types)) == 1:
+        if len(self.source_types) == 1:
             return self.source_annotations[0].type
-        elif len(set([ann.parent_type for ann in self.source_annotations])) == 1:
+        elif len(self.source_parent_types) == 1:
+            subtypes = [x for x in self.source_annotations if (x.type != x.parent_type)]
+            top = max(subtypes, key=attrgetter('score'))
+            if top.score >= TYPE_THRESHOLD:
+                return top.type
             return self.source_annotations[0].parent_type
         return "UNKNOWN"
 
+    @type.setter
+    def type(self, t):
+        pass
+
+    @property
+    def score(self):
+        if len(self.source_types) == 1:
+            return max([x.score for x in self.source_annotations])
+        elif len(self.source_parent_types) == 1:
+            subtypes = [x for x in self.source_annotations if (x.type != x.parent_type)]
+            top_score = max([x.score for x in subtypes])
+            if top_score >= TYPE_THRESHOLD:
+                return top_score
+            return max([x.score for x in self.source_annotations])
+        return TYPE_THRESHOLD
+
+    @score.setter
+    def score(self, t):
+        pass
+
     def add_annotation(self, ann):
         if ann.empty():
             raise ValueError("new annotation cannot be empty")
-        if self.empty():
+        elif self.empty():
             self.text = ann.text
             self.start = ann.start
             self.end = ann.end
+            self.type_map = ann.type_map
+        elif (self.end < ann.start) or (self.start > ann.end):
+            raise ValueError("annotation text must overlap")
         else:
             if (self.start <= ann.start):
                 self.text = self.text + ann.text[(self.end-ann.start):]
             else:
                 self.text = ann.text + self.text[(ann.end-self.start):]
             self.start = min([self.start, ann.start])
             self.end = max([self.end, ann.end])
+            self.type_map = {**ann.type_map, **self.type_map}
```

Comment:
what happens in there is a shared key in ann.type_map and self.type_map?

Reply:
so I tested this like:
d1 = {'a': 2, 'b':2}
d2 = {'b':3, 'c':4}
d3 = {**d1, **d2}

and got a syntax error. What is this doing?

Reply:
....okay so I tried some more and got:
d3 = dict(d1, **d2)
d3
{'a': 2, 'c': 4, 'b': 3}

and have one conclusion and two question:
conclusion: the second arg (**self.type_map) will overwrite any duplicate keys from ann.type_map
questions:
is this acceptable?
Does the syntax you are using work? and if so why is it blowing up for me?

Reply:
> is this acceptable?

Short answer: Yes, I think so, but I'm open to changing it.
Long answer: With our current setup, this doesn't actually matter; we only have one possible type/parent mapping dict (for hutchner), so if it ever overwrites, it'll be with the exact same values.
Where this _might_ come into play _eventually_, is if we introduce more PHI systems, _and_ those systems have the same type labels as hutchner, except they mean different things (for example: 'WARD_NAME' labels from hutchner are a subtype of 'ADDRESS', but maybe in our other system they'd be a subtype 'NAME' or 'ID'). In that case, in our MergedAnnotation rollup, I think it makes more sense for the first Annotations brought in to take priority when conflicts arise.
That said, it's a case that (a) will _never_ arrive with our current systems, and (b) would be extremely niche even if we brought in a new system. So if we want to simplify it down to
`self.type_map = self.type_map or ann.type_map`
I wouldn't object.

> Does the syntax you are using work? and if so why is it blowing up for me?

Hmm, yea, that syntax definitely works for me. Do the unit tests work for you? Because those definitely call that code.

I think it's kinda a cool approach, but if it's causing issue, I'm totally fine with switching it to only pull in if the map is currently empty (`= self.type_map or ann.type_map`).

Reply:
I double-checked in another console (still python3), and it seems to be working for me:
```
>>> d1 = {"a":1, "b":2}
>>> d2 = {"a":9, "c":5}
>>> d3 = {**d1, **d2}
>>> d3
{'a': 9, 'c': 5, 'b': 2}
```

But to move this forward, I'm just gonna change this to `self.type_map = self.type_map or ann.type_map`. We can explore more complex options if/when we add additional PHI detection systems into the mix.